File: src/main/java/alfheim/common/block/tile/sub/SubTileManaVoid.java
package alfheim.common.block.tile.sub;

import java.util.List;

import alexsocol.asjlib.ASJUtilities;
import alexsocol.asjlib.math.Vector3;
import alfheim.api.block.tile.SubTileEntity;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import vazkii.botania.api.mana.ManaItemHandler;
import vazkii.botania.common.Botania;

public class SubTileManaVoid extends SubTileEntity {

	public static final String TAG_MANA = "mana";
	public static int radius = 10;
	public int mana;
	Vector3 v = new Vector3();
	
	@Override
	public void update() {
		if (mana >= 120000) {
			for (EntityPlayer player : allAround(EntityPlayer.class, radius)) {
				radius = 50;
				for (int i = 0; i < 100; i++) performEffect(player);
			}
			
			superTile.getWorldObj().createExplosion(null, superTile.xCoord, superTile.yCoord, superTile.zCoord, radius = 10, false);
			
			for (int i = 0; i < 128; i++) {
				v.rand().sub(0.5).normalize().mul(Math.random() * 0.1);
				Botania.proxy.wispFX(superTile.getWorldObj(), superTile.xCoord + 0.5, superTile.yCoord + 0.5, superTile.zCoord + 0.5, 0.01F, 0.75F, 1, 0.25F, (float) v.x, (float) v.y, (float) v.z, 2);
			}
			
			mana = 0;
		}
	}
	
	@Override
	public List<Object> getTargets() {
		return allAroundRaw(EntityPlayer.class, radius);
	}

	@Override
	public void performEffect(Object target) {
		if (target == null || !(target instanceof EntityPlayer)) return;
		EntityPlayer player = (EntityPlayer) target;
		
		int m = ManaItemHandler.requestMana(new ItemStack(Blocks.stone), player, 100, true);
		if (m > 0) {
			mana += m;
			
			boolean flag = false;
			if (!ASJUtilities.isServer()) flag = Minecraft.getMinecraft().thePlayer != player;
			
			double l = v.set(superTile).add(0.5).sub(player.posX, player.posY + (flag ? 1 : -0.62), player.posZ).length();
			v.normalize().mul(l / 40);
			Botania.proxy.wispFX(superTile.getWorldObj(), player.posX, player.posY + (flag ? 1 : -0.62), player.posZ, 0.01F, 0.75F, 1, radius / 40F, (float) v.x, (float) v.y, (float) v.z, 2);
		}
	}

	@Override
	public int typeBits() {
		return MANA;
	}
	
	@Override
	public void writeCustomNBT(NBTTagCompound cmp) {
		super.writeCustomNBT(cmp);
		cmp.setInteger(TAG_MANA, mana);
	}
	
	@Override
	public void readCustomNBT(NBTTagCompound cmp) {
		super.writeCustomNBT(cmp);
		mana = cmp.getInteger(TAG_MANA);
	}
}

File: src/main/java/alfheim/common/block/tile/sub/anomaly/SubTileManaVoid.kt
package alfheim.common.block.tile.sub.anomaly

import alexsocol.asjlib.ASJUtilities
import alexsocol.asjlib.math.Vector3
import alfheim.api.block.tile.SubTileEntity
import alfheim.client.core.util.mc
import alfheim.client.render.world.VisualEffectHandlerClient
import alfheim.common.core.handler.VisualEffectHandler
import alfheim.common.core.util.*
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Blocks
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import vazkii.botania.api.mana.ManaItemHandler

class SubTileManaVoid: SubTileEntity() {
	var mana: Int = 0
	internal val v = Vector3()
	
	override val targets: List<Any>
		get() = if (inWG()) EMPTY_LIST else allAroundRaw(EntityPlayer::class.java, radius.D)
	
	override val strip: Int
		get() = 3
	
	override val rarity: EnumAnomalityRarity
		get() = EnumAnomalityRarity.COMMON
	
	public override fun update() {
		if (mana >= 120000) {
			for (player in allAround(EntityPlayer::class.java, radius.D)) {
				radius = 50
				for (i in 0..99) performEffect(player)
			}
			
			radius = 10
			worldObj.createExplosion(null, x().D, y().D, z().D, radius.F, false)
			
			VisualEffectHandler.sendPacket(VisualEffectHandlerClient.VisualEffects.MANABURST, worldObj.provider.dimensionId, x().D, y().D, z().D)
			
			mana = 0
		}
	}
	
	override fun performEffect(target: Any) {
		if (target !is EntityPlayer) return
		
		val m = ManaItemHandler.requestMana(ItemStack(Blocks.stone), target, 100, true)
		if (m > 0) {
			mana += m
			
			var flag = false
			if (!ASJUtilities.isServer) flag = mc.thePlayer !== target
			
			VisualEffectHandler.sendPacket(VisualEffectHandlerClient.VisualEffects.MANAVOID, worldObj.provider.dimensionId, x().D, y().D, z().D, target.posX, target.posY + if (flag) 1.0 else -0.62, target.posZ)
		}
	}
	
	override fun typeBits() = MANA
	
	override fun writeCustomNBT(cmp: NBTTagCompound) {
		super.writeCustomNBT(cmp)
		cmp.setInteger(TAG_MANA, mana)
	}
	
	override fun readCustomNBT(cmp: NBTTagCompound) {
		super.writeCustomNBT(cmp)
		mana = cmp.getInteger(TAG_MANA)
	}
	
	companion object {
		const val TAG_MANA = "mana"
		var radius = 10
	}
}
